### ISSUE TITLE
Stack Frame Walking / Stack Information Retrieval

### ISSUE THREAD

**Issue by Shinmera** [clasp-developers/clasp#584]
This is a feature request / informational issue in order to get support for Clasp into the debugger implementation support library [Dissect](https://github.com/Shinmera/dissect). In order to function, Dissect needs:

* [ ] A way to traverse the stack downwards from the current point, frame by frame.

For each frame it needs the following information:

* [ ] What function was called on this frame
* [ ] Which arguments were used in this frame
* [ ] (Optional) What is the source form of the called function
* [ ] (Optional) What is the source location in a file of the called function
* [ ] (Optional) How to evaluate an expression in the frame

For each restart it needs the following information:

* [ ] The restart's report function
* [ ] The restart's body function
* [ ] The restart's test function
* [ ] The restart's interactive function

For more information see the [current implementation](https://github.com/Shinmera/dissect/tree/master/backend).

Please edit my post with the appropriate Clasp names for the required features if or when they exist.

**Comment by Bike** [comment]
Restart information is available via `ext:restart-report-function`, `ext:restart-function`, `ext:restart-test-function`, and `ext:restart-interactive-function` respectively, as of db7ceacd661283643966d5ca78c3d3298ba88a7e. There's also `ext:restart-associated-conditions` since Dissect has that for SBCL.

I'll write a Dissect implementation myself once we stabilize a backtrace interface.

**Comment by Shinmera** [comment]
Thank you very much in advance!

**Comment by Bike** [comment]
Filed as https://github.com/Shinmera/dissect/pull/10

**Comment by Bike** [comment]
Evaluating in frame... that's kind of weird to me conceptually. Lots of fiddly bits.

For SLDB, what we do right now is just eval `(let (...variable bindings...) form)`. We don't even do that correctly, since the frame variable bindings aren't complete right now. Even if they were complete, we'd have to know which variables are actually in place. I don't think even SBCL knows that, they just dump multiple variables with the same name, according to the manual.

The dynamic environment is weird too. Presumably the form should be evaluated in the dynamic environment of the frame's call, except that afterward we need to restore the dynamic environment of the debugger.

**Comment by Shinmera** [comment]
Yeah. That part in the very least is definitely something Dissect would only support doing while sitting on top of the stack. The exact semantics of it are probably pretty different from  implementation to implementation, too.
